File: tools/sources/rust/keywords.rs
//! Every word in rust/config/keywords.txt used in a naming position.
//!
//! Rust reserves all but `union` (contextual), so bare use is almost
//! impossible. Raw identifiers (`r#type`) cover 48 of the 52, and capitalized
//! or upper-case variants cover the rest -- ordinary Rust identifiers that the
//! symbol filter folds onto a keyword, because it lowercases before comparing.
//!
//! `self` and `Self` collapse to the same variant, so the case-variant
//! sections carry one of each pair.
//!
//! Note: the list entry `Self` carries a capital and is compared with strcmp
//! against an already lowercased symbol, so it can never match anything.
//!
//! All identifiers are two characters or longer, so nothing here is dropped on
//! minimum length rather than on keyword-ness.

// Case variants are deliberately not snake_case: the point is to collide with
// the keyword list after lowercasing.
#![allow(non_snake_case)]

// Upper-case variants as consts.
pub const AS: i32 = 1;
pub const ASYNC: i32 = 2;
pub const AWAIT: i32 = 3;
pub const BREAK: i32 = 4;
pub const CONST: i32 = 5;
pub const CONTINUE: i32 = 6;
pub const CRATE: i32 = 7;
pub const DYN: i32 = 8;
pub const ELSE: i32 = 9;
pub const ENUM: i32 = 10;
pub const EXTERN: i32 = 11;
pub const FALSE: i32 = 12;
pub const FN: i32 = 13;
pub const FOR: i32 = 14;
pub const IF: i32 = 15;
pub const IMPL: i32 = 16;
pub const IN: i32 = 17;
pub const LET: i32 = 18;
pub const LOOP: i32 = 19;
pub const MATCH: i32 = 20;
pub const MOD: i32 = 21;
pub const MOVE: i32 = 22;
pub const MUT: i32 = 23;
pub const PUB: i32 = 24;
pub const REF: i32 = 25;
pub const RETURN: i32 = 26;
pub const SELF: i32 = 27;
pub const STATIC: i32 = 28;
pub const STRUCT: i32 = 29;
pub const SUPER: i32 = 30;
pub const TRAIT: i32 = 31;
pub const TRUE: i32 = 32;
pub const TYPE: i32 = 33;
pub const UNSAFE: i32 = 34;
pub const USE: i32 = 35;
pub const WHERE: i32 = 36;
pub const WHILE: i32 = 37;
pub const ABSTRACT: i32 = 38;
pub const BECOME: i32 = 39;
pub const BOX: i32 = 40;
pub const DO: i32 = 41;
pub const FINAL: i32 = 42;
pub const MACRO: i32 = 43;
pub const OVERRIDE: i32 = 44;
pub const PRIV: i32 = 45;
pub const TYPEOF: i32 = 46;
pub const UNSIZED: i32 = 47;
pub const VIRTUAL: i32 = 48;
pub const YIELD: i32 = 49;
pub const TRY: i32 = 50;
pub const UNION: i32 = 51;

// Capitalized variants as struct names.
pub struct AsHolder { pub mem: i32 }
pub struct AsyncHolder { pub mem: i32 }
pub struct AwaitHolder { pub mem: i32 }
pub struct BreakHolder { pub mem: i32 }
pub struct ConstHolder { pub mem: i32 }
pub struct ContinueHolder { pub mem: i32 }
pub struct CrateHolder { pub mem: i32 }
pub struct DynHolder { pub mem: i32 }
pub struct ElseHolder { pub mem: i32 }
pub struct EnumHolder { pub mem: i32 }
pub struct ExternHolder { pub mem: i32 }
pub struct FalseHolder { pub mem: i32 }
pub struct FnHolder { pub mem: i32 }
pub struct ForHolder { pub mem: i32 }
pub struct IfHolder { pub mem: i32 }
pub struct ImplHolder { pub mem: i32 }
pub struct InHolder { pub mem: i32 }
pub struct LetHolder { pub mem: i32 }
pub struct LoopHolder { pub mem: i32 }
pub struct MatchHolder { pub mem: i32 }
pub struct ModHolder { pub mem: i32 }
pub struct MoveHolder { pub mem: i32 }
pub struct MutHolder { pub mem: i32 }
pub struct PubHolder { pub mem: i32 }
pub struct RefHolder { pub mem: i32 }
pub struct ReturnHolder { pub mem: i32 }
pub struct SelfHolder { pub mem: i32 }
pub struct StaticHolder { pub mem: i32 }
pub struct StructHolder { pub mem: i32 }
pub struct SuperHolder { pub mem: i32 }
pub struct TraitHolder { pub mem: i32 }
pub struct TrueHolder { pub mem: i32 }
pub struct TypeHolder { pub mem: i32 }
pub struct UnsafeHolder { pub mem: i32 }
pub struct UseHolder { pub mem: i32 }
pub struct WhereHolder { pub mem: i32 }
pub struct WhileHolder { pub mem: i32 }
pub struct AbstractHolder { pub mem: i32 }
pub struct BecomeHolder { pub mem: i32 }
pub struct BoxHolder { pub mem: i32 }
pub struct DoHolder { pub mem: i32 }
pub struct FinalHolder { pub mem: i32 }
pub struct MacroHolder { pub mem: i32 }
pub struct OverrideHolder { pub mem: i32 }
pub struct PrivHolder { pub mem: i32 }
pub struct TypeofHolder { pub mem: i32 }
pub struct UnsizedHolder { pub mem: i32 }
pub struct VirtualHolder { pub mem: i32 }
pub struct YieldHolder { pub mem: i32 }
pub struct TryHolder { pub mem: i32 }
pub struct UnionHolder { pub mem: i32 }

// Capitalized variants as struct fields.
pub struct KeywordHolder {
    pub As: i32,
    pub Async: i32,
    pub Await: i32,
    pub Break: i32,
    pub Const: i32,
    pub Continue: i32,
    pub Crate: i32,
    pub Dyn: i32,
    pub Else: i32,
    pub Enum: i32,
    pub Extern: i32,
    pub False: i32,
    pub Fn: i32,
    pub For: i32,
    pub If: i32,
    pub Impl: i32,
    pub In: i32,
    pub Let: i32,
    pub Loop: i32,
    pub Match: i32,
    pub Mod: i32,
    pub Move: i32,
    pub Mut: i32,
    pub Pub: i32,
    pub Ref: i32,
    pub Return: i32,
    pub Static: i32,
    pub Struct: i32,
    pub Super: i32,
    pub Trait: i32,
    pub True: i32,
    pub Type: i32,
    pub Unsafe: i32,
    pub Use: i32,
    pub Where: i32,
    pub While: i32,
    pub Abstract: i32,
    pub Become: i32,
    pub Box: i32,
    pub Do: i32,
    pub Final: i32,
    pub Macro: i32,
    pub Override: i32,
    pub Priv: i32,
    pub Typeof: i32,
    pub Unsized: i32,
    pub Virtual: i32,
    pub Yield: i32,
    pub Try: i32,
    pub Union: i32,
}

// Capitalized variants as function names.
pub fn AsFn() -> i32 { 1 }
pub fn AsyncFn() -> i32 { 1 }
pub fn AwaitFn() -> i32 { 1 }
pub fn BreakFn() -> i32 { 1 }
pub fn ConstFn() -> i32 { 1 }
pub fn ContinueFn() -> i32 { 1 }
pub fn CrateFn() -> i32 { 1 }
pub fn DynFn() -> i32 { 1 }
pub fn ElseFn() -> i32 { 1 }
pub fn EnumFn() -> i32 { 1 }
pub fn ExternFn() -> i32 { 1 }
pub fn FalseFn() -> i32 { 1 }
pub fn FnFn() -> i32 { 1 }
pub fn ForFn() -> i32 { 1 }
pub fn IfFn() -> i32 { 1 }
pub fn ImplFn() -> i32 { 1 }
pub fn InFn() -> i32 { 1 }
pub fn LetFn() -> i32 { 1 }
pub fn LoopFn() -> i32 { 1 }
pub fn MatchFn() -> i32 { 1 }
pub fn ModFn() -> i32 { 1 }
pub fn MoveFn() -> i32 { 1 }
pub fn MutFn() -> i32 { 1 }
pub fn PubFn() -> i32 { 1 }
pub fn RefFn() -> i32 { 1 }
pub fn ReturnFn() -> i32 { 1 }
pub fn StaticFn() -> i32 { 1 }
pub fn StructFn() -> i32 { 1 }
pub fn SuperFn() -> i32 { 1 }
pub fn TraitFn() -> i32 { 1 }
pub fn TrueFn() -> i32 { 1 }
pub fn TypeFn() -> i32 { 1 }
pub fn UnsafeFn() -> i32 { 1 }
pub fn UseFn() -> i32 { 1 }
pub fn WhereFn() -> i32 { 1 }
pub fn WhileFn() -> i32 { 1 }
pub fn AbstractFn() -> i32 { 1 }
pub fn BecomeFn() -> i32 { 1 }
pub fn BoxFn() -> i32 { 1 }
pub fn DoFn() -> i32 { 1 }
pub fn FinalFn() -> i32 { 1 }
pub fn MacroFn() -> i32 { 1 }
pub fn OverrideFn() -> i32 { 1 }
pub fn PrivFn() -> i32 { 1 }
pub fn TypeofFn() -> i32 { 1 }
pub fn UnsizedFn() -> i32 { 1 }
pub fn VirtualFn() -> i32 { 1 }
pub fn YieldFn() -> i32 { 1 }
pub fn TryFn() -> i32 { 1 }
pub fn UnionFn() -> i32 { 1 }

// Raw identifiers as let bindings: the keyword itself, escaped.
pub fn raw_bindings() -> i32 {
    let r#as = 1;
    let r#async = 1;
    let r#await = 1;
    let r#break = 1;
    let r#const = 1;
    let r#continue = 1;
    let r#dyn = 1;
    let r#else = 1;
    let r#enum = 1;
    let r#extern = 1;
    let r#false = 1;
    let r#fn = 1;
    let r#for = 1;
    let r#if = 1;
    let r#impl = 1;
    let r#in = 1;
    let r#let = 1;
    let r#loop = 1;
    let r#match = 1;
    let r#mod = 1;
    let r#move = 1;
    let r#mut = 1;
    let r#pub = 1;
    let r#ref = 1;
    let r#return = 1;
    let r#static = 1;
    let r#struct = 1;
    let r#trait = 1;
    let r#true = 1;
    let r#type = 1;
    let r#unsafe = 1;
    let r#use = 1;
    let r#where = 1;
    let r#while = 1;
    let r#abstract = 1;
    let r#become = 1;
    let r#box = 1;
    let r#do = 1;
    let r#final = 1;
    let r#macro = 1;
    let r#override = 1;
    let r#priv = 1;
    let r#typeof = 1;
    let r#unsized = 1;
    let r#virtual = 1;
    let r#yield = 1;
    let r#try = 1;
    let r#union = 1;
    r#as + r#async + r#await + r#break + r#const + r#continue + r#dyn + r#else + r#enum + r#extern + r#false + r#fn + r#for + r#if + r#impl + r#in + r#let + r#loop + r#match + r#mod + r#move + r#mut + r#pub + r#ref + r#return + r#static + r#struct + r#trait + r#true + r#type + r#unsafe + r#use + r#where + r#while + r#abstract + r#become + r#box + r#do + r#final + r#macro + r#override + r#priv + r#typeof + r#unsized + r#virtual + r#yield + r#try + r#union
}

// `union` is contextual, so it is legal bare.
pub fn union() -> i32 { 1 }

File: tools/sources/php/keywords.php
<?php

/*
 * Every word in php/config/keywords.txt used in a naming position.
 *
 * PHP allows reserved words as variable, property, method and class-constant
 * names; only class/function/trait/interface declaration names are barred.
 * Case variants are included because the symbol filter lowercases before it
 * compares, so `$For` collides with `for` even though PHP keeps them apart.
 *
 * All identifiers are two characters or longer, so nothing here is dropped
 * on minimum length rather than on keyword-ness.
 *
 * Position coverage was verified with php -l; the only word PHP rejects is
 * `class` as a class constant (reserved for ::class fetching).
 */

namespace App\Keywords;

// Keywords as local variables.
$abstract = 1;
$and = 2;
$array = 3;
$as = 4;
$break = 5;
$callable = 6;
$case = 7;
$catch = 8;
$class = 9;
$clone = 10;
$const = 11;
$continue = 12;
$declare = 13;
$default = 14;
$die = 15;
$do = 16;
$echo = 17;
$else = 18;
$elseif = 19;
$empty = 20;
$enddeclare = 21;
$endfor = 22;
$endforeach = 23;
$endif = 24;
$endswitch = 25;
$endwhile = 26;
$eval = 27;
$exit = 28;
$extends = 29;
$final = 30;
$finally = 31;
$fn = 32;
$for = 33;
$foreach = 34;
$function = 35;
$global = 36;
$goto = 37;
$if = 38;
$implements = 39;
$include = 40;
$include_once = 41;
$instanceof = 42;
$insteadof = 43;
$interface = 44;
$isset = 45;
$list = 46;
$match = 47;
$namespace = 48;
$new = 49;
$or = 50;
$print = 51;
$private = 52;
$protected = 53;
$public = 54;
$readonly = 55;
$require = 56;
$require_once = 57;
$return = 58;
$static = 59;
$switch = 60;
$throw = 61;
$trait = 62;
$try = 63;
$unset = 64;
$use = 65;
$var = 66;
$while = 67;
$xor = 68;
$yield = 69;

// Case variants: legal PHP identifiers the filter folds onto keywords.
$Abstract = 'abstract';
$And = 'and';
$Array = 'array';
$As = 'as';
$Break = 'break';
$Callable = 'callable';
$Case = 'case';
$Catch = 'catch';
$Class = 'class';
$Clone = 'clone';
$Const = 'const';
$Continue = 'continue';
$Declare = 'declare';
$Default = 'default';
$Die = 'die';
$Do = 'do';
$Echo = 'echo';
$Else = 'else';
$Elseif = 'elseif';
$Empty = 'empty';
$Enddeclare = 'enddeclare';
$Endfor = 'endfor';
$Endforeach = 'endforeach';
$Endif = 'endif';
$Endswitch = 'endswitch';
$Endwhile = 'endwhile';
$Eval = 'eval';
$Exit = 'exit';
$Extends = 'extends';
$Final = 'final';
$Finally = 'finally';
$Fn = 'fn';
$For = 'for';
$Foreach = 'foreach';
$Function = 'function';
$Global = 'global';
$Goto = 'goto';
$If = 'if';
$Implements = 'implements';
$Include = 'include';
$Include_once = 'include_once';
$Instanceof = 'instanceof';
$Insteadof = 'insteadof';
$Interface = 'interface';
$Isset = 'isset';
$List = 'list';
$Match = 'match';
$Namespace = 'namespace';
$New = 'new';
$Or = 'or';
$Print = 'print';
$Private = 'private';
$Protected = 'protected';
$Public = 'public';
$Readonly = 'readonly';
$Require = 'require';
$Require_once = 'require_once';
$Return = 'return';
$Static = 'static';
$Switch = 'switch';
$Throw = 'throw';
$Trait = 'trait';
$Try = 'try';
$Unset = 'unset';
$Use = 'use';
$Var = 'var';
$While = 'while';
$Xor = 'xor';
$Yield = 'yield';

class KeywordHolder
{
    // Keywords as class constants (UPPER case variants).
    public const ABSTRACT = 'abstract';
    public const AND = 'and';
    public const ARRAY = 'array';
    public const AS = 'as';
    public const BREAK = 'break';
    public const CALLABLE = 'callable';
    public const CASE = 'case';
    public const CATCH = 'catch';
    public const CLONE = 'clone';
    public const CONST = 'const';
    public const CONTINUE = 'continue';
    public const DECLARE = 'declare';
    public const DEFAULT = 'default';
    public const DIE = 'die';
    public const DO = 'do';
    public const ECHO = 'echo';
    public const ELSE = 'else';
    public const ELSEIF = 'elseif';
    public const EMPTY = 'empty';
    public const ENDDECLARE = 'enddeclare';
    public const ENDFOR = 'endfor';
    public const ENDFOREACH = 'endforeach';
    public const ENDIF = 'endif';
    public const ENDSWITCH = 'endswitch';
    public const ENDWHILE = 'endwhile';
    public const EVAL = 'eval';
    public const EXIT = 'exit';
    public const EXTENDS = 'extends';
    public const FINAL = 'final';
    public const FINALLY = 'finally';
    public const FN = 'fn';
    public const FOR = 'for';
    public const FOREACH = 'foreach';
    public const FUNCTION = 'function';
    public const GLOBAL = 'global';
    public const GOTO = 'goto';
    public const IF = 'if';
    public const IMPLEMENTS = 'implements';
    public const INCLUDE = 'include';
    public const INCLUDE_ONCE = 'include_once';
    public const INSTANCEOF = 'instanceof';
    public const INSTEADOF = 'insteadof';
    public const INTERFACE = 'interface';
    public const ISSET = 'isset';
    public const LIST = 'list';
    public const MATCH = 'match';
    public const NAMESPACE = 'namespace';
    public const NEW = 'new';
    public const OR = 'or';
    public const PRINT = 'print';
    public const PRIVATE = 'private';
    public const PROTECTED = 'protected';
    public const PUBLIC = 'public';
    public const READONLY = 'readonly';
    public const REQUIRE = 'require';
    public const REQUIRE_ONCE = 'require_once';
    public const RETURN = 'return';
    public const STATIC = 'static';
    public const SWITCH = 'switch';
    public const THROW = 'throw';
    public const TRAIT = 'trait';
    public const TRY = 'try';
    public const UNSET = 'unset';
    public const USE = 'use';
    public const VAR = 'var';
    public const WHILE = 'while';
    public const XOR = 'xor';
    public const YIELD = 'yield';

    // Keywords as properties.
    public $abstract = null;
    public $and = null;
    public $array = null;
    public $as = null;
    public $break = null;
    public $callable = null;
    public $case = null;
    public $catch = null;
    public $class = null;
    public $clone = null;
    public $const = null;
    public $continue = null;
    public $declare = null;
    public $default = null;
    public $die = null;
    public $do = null;
    public $echo = null;
    public $else = null;
    public $elseif = null;
    public $empty = null;
    public $enddeclare = null;
    public $endfor = null;
    public $endforeach = null;
    public $endif = null;
    public $endswitch = null;
    public $endwhile = null;
    public $eval = null;
    public $exit = null;
    public $extends = null;
    public $final = null;
    public $finally = null;
    public $fn = null;
    public $for = null;
    public $foreach = null;
    public $function = null;
    public $global = null;
    public $goto = null;
    public $if = null;
    public $implements = null;
    public $include = null;
    public $include_once = null;
    public $instanceof = null;
    public $insteadof = null;
    public $interface = null;
    public $isset = null;
    public $list = null;
    public $match = null;
    public $namespace = null;
    public $new = null;
    public $or = null;
    public $print = null;
    public $private = null;
    public $protected = null;
    public $public = null;
    public $readonly = null;
    public $require = null;
    public $require_once = null;
    public $return = null;
    public $static = null;
    public $switch = null;
    public $throw = null;
    public $trait = null;
    public $try = null;
    public $unset = null;
    public $use = null;
    public $var = null;
    public $while = null;
    public $xor = null;
    public $yield = null;

    // Keywords as method names.
    public function abstract() { return $this->abstract; }
    public function and() { return $this->and; }
    public function array() { return $this->array; }
    public function as() { return $this->as; }
    public function break() { return $this->break; }
    public function callable() { return $this->callable; }
    public function case() { return $this->case; }
    public function catch() { return $this->catch; }
    public function class() { return $this->class; }
    public function clone() { return $this->clone; }
    public function const() { return $this->const; }
    public function continue() { return $this->continue; }
    public function declare() { return $this->declare; }
    public function default() { return $this->default; }
    public function die() { return $this->die; }
    public function do() { return $this->do; }
    public function echo() { return $this->echo; }
    public function else() { return $this->else; }
    public function elseif() { return $this->elseif; }
    public function empty() { return $this->empty; }
    public function enddeclare() { return $this->enddeclare; }
    public function endfor() { return $this->endfor; }
    public function endforeach() { return $this->endforeach; }
    public function endif() { return $this->endif; }
    public function endswitch() { return $this->endswitch; }
    public function endwhile() { return $this->endwhile; }
    public function eval() { return $this->eval; }
    public function exit() { return $this->exit; }
    public function extends() { return $this->extends; }
    public function final() { return $this->final; }
    public function finally() { return $this->finally; }
    public function fn() { return $this->fn; }
    public function for() { return $this->for; }
    public function foreach() { return $this->foreach; }
    public function function() { return $this->function; }
    public function global() { return $this->global; }
    public function goto() { return $this->goto; }
    public function if() { return $this->if; }
    public function implements() { return $this->implements; }
    public function include() { return $this->include; }
    public function include_once() { return $this->include_once; }
    public function instanceof() { return $this->instanceof; }
    public function insteadof() { return $this->insteadof; }
    public function interface() { return $this->interface; }
    public function isset() { return $this->isset; }
    public function list() { return $this->list; }
    public function match() { return $this->match; }
    public function namespace() { return $this->namespace; }
    public function new() { return $this->new; }
    public function or() { return $this->or; }
    public function print() { return $this->print; }
    public function private() { return $this->private; }
    public function protected() { return $this->protected; }
    public function public() { return $this->public; }
    public function readonly() { return $this->readonly; }
    public function require() { return $this->require; }
    public function require_once() { return $this->require_once; }
    public function return() { return $this->return; }
    public function static() { return $this->static; }
    public function switch() { return $this->switch; }
    public function throw() { return $this->throw; }
    public function trait() { return $this->trait; }
    public function try() { return $this->try; }
    public function unset() { return $this->unset; }
    public function use() { return $this->use; }
    public function var() { return $this->var; }
    public function while() { return $this->while; }
    public function xor() { return $this->xor; }
    public function yield() { return $this->yield; }
}

// Keywords as function arguments and named-argument labels.
function describe(string $class_name, int $for = 0, bool $match = false)
{
    return $class_name . $for . $match;
}

